### PR TITLE
changed u_int32_t to uint32_t, to compile on Alpine Linux (musl)

### PR DESCRIPTION
**Pull request by ehudkaldor**
Alpine is using musl instead of glibc, so it might be a bit more strict. from what I could find, the REAL definition in C call it uint32_t (see [https://stackoverflow.com/questions/5163835/what-is-u-int32-t](https://stackoverflow.com/questions/5163835/what-is-u-int32-t)).
it seems to compile under Ubuntu as well, after the change

### CODE REVIEW

**Comment by ehudkaldor** [comment]
you are right, and I do apologize. I am not familiar with this technique of building for specific platforms. I researched why it does not install on Alpine linux, and found the uint vs. u_int issue, replaced all and it worked. 
I thought this was a bug and thought to submit a patch. and removing cffi was only in response to the Travis autobuild saying it failed because it cannot find install cffi. Again, installing cffi on Alpine using pip fails, and only the one in the Alpine repo (py-cffi) worked.

So, how can I get it to compile/install on Alpine? Alpine uses `musl` instead of `glibc`, but sys.platform still reports it as "linux2".

**Comment by etrepum** [comment]
This sounds like a problem that alpine porters must run into often, maybe you can find an answer in their community?

**Comment by ehudkaldor** [comment]
i've been on the tail of this issue for some time. here is a question I posted on stack overflow in Jan, when I thought it was a lib32 issue. no answer, not even a comment:
https://stackoverflow.com/questions/48387759/how-to-use-32bit-python-libs-on-alpine-linux
the only way I could get it to build is by making that change. I don't know, however, which of your platform branches Alpine falls into.

all those IFDEF you have there for SUN and BSD - how are they recognized? is it part of cffi that knows that? maybe it will also know how to recognize musl.

just for reference, here is `pip install xattr` output on Alpine. OS packages installed: python, py-pip, musl-dev, libc-dev, gcc:

```
/ # pip install xattr
Collecting xattr
  Using cached https://files.pythonhosted.org/packages/ea/d8/d4e412bd0501ac3f41640595d32b6d8fd6640452436b228e081c0557db53/xattr-0.9.3.tar.gz
Requirement already satisfied: cffi>=1.0.0 in /usr/lib/python2.7/site-packages (from xattr)
Requirement already satisfied: pycparser in /usr/lib/python2.7/site-packages (from cffi>=1.0.0->xattr)
Installing collected packages: xattr
  Running setup.py install for xattr ... error
    Complete output from command /usr/bin/python2 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-JSqj9x/xattr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-swg6Ut-record/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-2.7
    creating build/lib.linux-x86_64-2.7/xattr
    copying xattr/tool.py -> build/lib.linux-x86_64-2.7/xattr
    copying xattr/lib_build.py -> build/lib.linux-x86_64-2.7/xattr
    copying xattr/lib.py -> build/lib.linux-x86_64-2.7/xattr
    copying xattr/pyxattr_compat.py -> build/lib.linux-x86_64-2.7/xattr
    copying xattr/__init__.py -> build/lib.linux-x86_64-2.7/xattr
    copying xattr/compat.py -> build/lib.linux-x86_64-2.7/xattr
    running build_ext
    generating cffi module 'build/temp.linux-x86_64-2.7/_lib.c'
    creating build/temp.linux-x86_64-2.7
    building '_lib' extension
    creating build/temp.linux-x86_64-2.7/build
    creating build/temp.linux-x86_64-2.7/build/temp.linux-x86_64-2.7
    gcc -fno-strict-aliasing -Os -fomit-frame-pointer -g -DNDEBUG -Os -fomit-frame-pointer -g -DTHREAD_STACK_SIZE=0x100000 -fPIC -I/usr/include/python2.7 -c build/temp.linux-x86_64-2.7/_lib.c -o build/temp.linux-x86_64-2.7/build/temp.linux-x86_64-2.7/_lib.o
    build/temp.linux-x86_64-2.7/_lib.c:869:94: error: unknown type name 'u_int32_t'
     static ssize_t xattr_getxattr(const char *path, const char *name, void *value, ssize_t size, u_int32_t position, int options) {
                                                                                                  ^~~~~~~~~
    build/temp.linux-x86_64-2.7/_lib.c:880:94: error: unknown type name 'u_int32_t'
     static ssize_t xattr_setxattr(const char *path, const char *name, void *value, ssize_t size, u_int32_t position, int options) {
                                                                                                  ^~~~~~~~~
    build/temp.linux-x86_64-2.7/_lib.c:924:85: error: unknown type name 'u_int32_t'
     static ssize_t xattr_fgetxattr(int fd, const char *name, void *value, ssize_t size, u_int32_t position, int options) {
                                                                                         ^~~~~~~~~
    build/temp.linux-x86_64-2.7/_lib.c:935:85: error: unknown type name 'u_int32_t'
     static ssize_t xattr_fsetxattr(int fd, const char *name, void *value, ssize_t size, u_int32_t position, int options) {
                                                                                         ^~~~~~~~~
    build/temp.linux-x86_64-2.7/_lib.c: In function '_cffi_d_xattr_fgetxattr':
    build/temp.linux-x86_64-2.7/_lib.c:1048:10: warning: implicit declaration of function 'xattr_fgetxattr' [-Wimplicit-function-declaration]
       return xattr_fgetxattr(x0, x1, x2, x3, x4, x5);
              ^~~~~~~~~~~~~~~
    build/temp.linux-x86_64-2.7/_lib.c: In function '_cffi_d_xattr_fsetxattr':
    build/temp.linux-x86_64-2.7/_lib.c:1235:10: warning: implicit declaration of function 'xattr_fsetxattr' [-Wimplicit-function-declaration]
       return xattr_fsetxattr(x0, x1, x2, x3, x4, x5);
              ^~~~~~~~~~~~~~~
    build/temp.linux-x86_64-2.7/_lib.c: In function '_cffi_d_xattr_getxattr':
    build/temp.linux-x86_64-2.7/_lib.c:1312:10: warning: implicit declaration of function 'xattr_getxattr' [-Wimplicit-function-declaration]
       return xattr_getxattr(x0, x1, x2, x3, x4, x5);
              ^~~~~~~~~~~~~~
    build/temp.linux-x86_64-2.7/_lib.c: In function '_cffi_d_xattr_setxattr':
    build/temp.linux-x86_64-2.7/_lib.c:1520:10: warning: implicit declaration of function 'xattr_setxattr' [-Wimplicit-function-declaration]
       return xattr_setxattr(x0, x1, x2, x3, x4, x5);
              ^~~~~~~~~~~~~~
    error: command 'gcc' failed with exit status 1
    
    ----------------------------------------
Command "/usr/bin/python2 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-JSqj9x/xattr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-swg6Ut-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-JSqj9x/xattr/
You are using pip version 9.0.1, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

**Comment by nyov** [comment]
The `cffi` is actually the 'c foreign function interface' library, which does compile C code, and those `ifdef`s are recognized by the C compiler.
Those `ifdef` magics are usually defined by the OS environment, e.g. FreeBSD [notes it here](https://www.freebsd.org/doc/en/books/porters-handbook/porting-versions.html) and a lot others are listed here: [https://sourceforge.net/p/predef/wiki/OperatingSystems/](https://sourceforge.net/p/predef/wiki/OperatingSystems/).
Also, don't apologize please for writing a patch! It's good to know there is an issue, and good work -- I'm just rumbling about unrelated changes, and you changed that back.
It _is_ a bug (or issue) and this is the right place to fix it. But yesterday it did look wrong to change it for _all_ the platforms.
Have you tried if my mentioned `ifndef` in the Linux section would work for you?


@etrepum, the most common solution I found from a casual websearch does seem to be switching the datatype to `uint_*_t`, as declared by C99 and common to Linux/System V, while `u_int_*_t` appears to be the BSD way (and by extension Mac OS X) of old, but a [current `sys/types.h` header in NetBSD](https://ftp.netbsd.org/pub/NetBSD/NetBSD-current/src/sys/sys/types.h) also seems to define `u_int_*_t` from `uint_*_t`.
```
typedef	uint8_t		u_int8_t;
typedef	uint16_t	u_int16_t;
typedef	uint32_t	u_int32_t;
typedef	uint64_t	u_int64_t;
```
and FreeBSD even seems to [call them deprecated](https://github.com/freebsd/freebsd/blob/8bff5fd9562a73b25b75a806911ab77253643ea1/sys/sys/types.h#L67-L70). OpenBSD also defines this C99 style before the BSD style.

All in all I can't be certain; though it seems recent platforms define both types (except Alpine/musli and XNU) and for Linux `uint_*_t` looks to be the right choice.
Except for XNU, most recent BSDs seem to have the C99 style also; but maybe I'm looking in the wrong place for XNU/OS X. I did not find anything on Solaris this shortly.

So today I'd feel reasonably safe to change to `uint_*_t` types and put a global
```
#ifndef u_int32_t
#define u_int32_t uint32_t
#endif
```
for Solaris, Mac OS and possibly any other holdovers.
What do you think? If you prefer that BSD style, the other way around works as well ofc.

https://ftp.netbsd.org/pub/NetBSD/NetBSD-current/src/sys/sys/types.h
https://github.com/freebsd/freebsd/blob/master/sys/sys/types.h
https://github.com/openbsd/src/blob/master/sys/sys/types.h
https://www.gnu.org/software/libc/manual/html_node/Integers.html
https://github.com/opensource-apple/xnu/blob/master/bsd/sys/types.h

**Comment by etrepum** [comment]
C99 style seems to be the right choice, thanks for doing all that research!

**Comment by nyov** [comment]
Ok, I'll try myself at a patch, unless you're at it already?

**Comment by ehudkaldor** [comment]
I just wanted to say you guys rock, and thank you. do you want me to open an issue for this?

**Comment by nyov** [comment]
@ehudkaldor I got sidetracked, and was now waiting on whether #70 gets accepted, because I based my patch on that.
But I opened a PR for it now (#71) so you can pull and test it. It's really quite similar to yours, so I made you the author of it, for the credit. And let me know if it works for you!

**Comment by etrepum** [comment]
I'd like to sort out why travis builds fail now before merging anything, I may have some time to look into it this weekend

**Comment by nyov** [comment]
Yeah would be nice if travis was fixed.
I looked at it, but that wrangling there to make it build python modules on OSX is a bit heavy to grasp.
FWIW, looks like all the builders fail at installing/importing pip, which tries to get `pyexpat` and fails to find it.

```
Last 10 log lines:

    import pip
[...]
    from pyexpat import *
```
The exception being the BUILD_SDIST target, which fails with an SSL error:
```
+python setup.py build_ext -i
Download error on https://pypi.python.org/simple/cffi/: [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:590) -- Some packages may not be found!
```

File: xattr/lib_build.py [hunk 1]
```diff
@@ -272,12 +272,12 @@
 #define XATTR_CREATE 0x1
 #define XATTR_REPLACE 0x2
 
-#ifndef u_int32_t
-#define u_int32_t uint32_t
+#ifndef uint32_t
+#define uint32_t uint32_t
```
Comment (alex) [inline xattr/lib_build.py]: This looks wrong to me; I suspect this whole section can be removed

Comment (nyov) [reply xattr/lib_build.py]: Yes, we wanted to replace this PR with #71. Could you review that one, perhaps? Thanks!

**Comment by nyov** [comment]
@etrepum is there an update with the CI issues?

**Comment by etrepum** [comment]
Unfortunately no. I tried a few things but couldn’t get it to work, and have been on vacation since.

**Comment by chrBrd** [comment]
While I think changing the types is probably the best way to go about this there is a much easier solution - see  #73.

**Comment by etrepum** [comment]
I think we can close this per #73

**Pull request closed without merging**